extensions/standard-processors/processors/InvokeHTTP.h

```diff
@@ -41,9 +42,60 @@
 #include "utils/Enum.h"
 #include "utils/RegexUtils.h"
 
+namespace org::apache::nifi::minifi::test {
+class HttpClientStoreTestAccessor;
+}
+
 namespace org::apache::nifi::minifi::processors {
 
 namespace invoke_http {
+class HttpClientStore {
+ public:
+  HttpClientStore(const size_t size, std::function<gsl::not_null<std::unique_ptr<minifi::http::HTTPClient>>(const std::string&)> create_client_function)
+      : max_size_(size),
+        create_client_function_(std::move(create_client_function)) {
+    clients_.resize(size);
+  }
+  HttpClientStore(const HttpClientStore&) = delete;
+  HttpClientStore& operator=(const HttpClientStore&) = delete;
+  HttpClientStore(HttpClientStore&&) = delete;
+  HttpClientStore& operator=(HttpClientStore&&) = delete;
+  ~HttpClientStore() = default;
+
+  class HttpClientWrapper {
+   public:
+    HttpClientWrapper(HttpClientStore& client_store, minifi::http::HTTPClient& client) : client_(client), client_store_(client_store) {
+    }
+    HttpClientWrapper(HttpClientWrapper&& src) = default;
+    HttpClientWrapper(const HttpClientWrapper&) = delete;
+    ~HttpClientWrapper() {
+      client_store_.returnClient(client_);
+    }
+
+    minifi::http::HTTPClient& get() const {
+      return client_;
+    }
+
+   private:
+    minifi::http::HTTPClient& client_;
+    HttpClientStore& client_store_;
+  };
+
+  [[nodiscard]] HttpClientWrapper getClient(const std::string& url);
+
+ private:
+  friend class ::org::apache::nifi::minifi::test::HttpClientStoreTestAccessor;
+  void returnClient(minifi::http::HTTPClient& client);
+
+  std::mutex clients_mutex_;
+  std::condition_variable cv_;
+  const size_t max_size_;
+  size_t clients_created_{0};
+  std::vector<std::pair<std::unique_ptr<minifi::http::HTTPClient>, bool>> clients_;
```

Comment:
since we are rotating a single element each time I think an `std::list` with `splice` would be clearer, also I think we should replace the `clients_created_` with `std::list::size` to act as a single source of truth

Reply:
Could you elaborate a bit on this, please? How would you use the `splice` with `std::list` in a way that would replace the rotate function?

We were also discussing using an `std::list` instead of the `std::vector` previously with @szaszm and I don't remember why we decided against it. @szaszm can you maybe remember?

Reply:
```
#include <list>
#include <algorithm>
#include <vector>
#include <iostream>

int main() {
  // std::rotate(
  //   clients_.begin() + gsl::narrow<int64_t>(i),
  //   clients_.begin() + gsl::narrow<int64_t>(i + 1),
  //   clients_.begin() + gsl::narrow<int64_t>(clients_created_)
  // )
  std::list<int> clients{1, 2, 3, 4};
  for (auto it = clients.begin(); it != clients.end(); ++it) {
    if (*it == 2) {
      clients.splice(clients.end(), clients, it);
      break;
    }
  }
  std::cout << "after rotateLeft\n";  // 1, 3, 4, 2
  for (auto& val : clients) {
    std::cout << val << std::endl;
  }

  clients = std::list<int>{1, 2, 3, 4};
  // rotateRight(
  //   clients_.begin() + last_unused + 1,
  //   clients_.begin() + gsl::narrow<int64_t>(i) + 1,
  //   1
  // );
  clients.splice(std::next(clients.begin(), 1), clients, std::next(clients.begin(), 2));

  std::cout << "after rotateRight\n";  // 1, 3, 2, 4
  for (auto& val : clients) {
    std::cout << val << std::endl;
  }
}
```

Reply:
I don't think it brings any performance benefits or things like that (I don't imagine we will have more than a handful of cached clients so the linear scanning works just fine), just that IMO it is a better fit to how we are actually using this container